docs/dev/core/new-metric-type.md

```diff
@@ -0,0 +1,212 @@
+
+A metric type implementation is defined in its own file under `glean-core/src/metrics/`, e.g. `glean-core/src/metrics/counter.rs` for a [Counter](/user/metrics/counter.md).
+
+Start by defining a structure:
+
+```rust,noplaypen
+#[derive(Debug)]
+pub struct CounterMetric {
+    meta: CommonMetricData
+}
+```
+
+Its implementation should have a way to create a new metric from the common metric data. It should be the same for all metric types.
+
+```rust,noplaypen
+impl CounterMetric {
+    pub fn new(meta: CommonMetricData) -> Self {
+        Self { meta }
+    }
+}
+```
+
+Implement each method. The first argument to accept should always be `glean: &Glean`, that is: a reference to the Glean object, used to access the storage:
+
+```rust,noplaypen
+impl CounterMetric { // same block as above
+    pub fn add(&self, glean: &Glean, amount: i32) {
+        // Always include this check!
+        if !self.meta.should_record() || !glean.is_upload_enabled() {
+            return;
+        }
+
+        // Do error handling here
+
+        glean
+            .storage()
+            .record_with(&self.meta, |old_value| match old_value {
+                Some(Metric::Counter(old_value)) => Metric::Counter(old_value + amount),
+                _ => Metric::Counter(amount),
+            })
+    }
+}
+```
+
+Use `glean.storage().record()` to record a fixed value or `glean.storage.record_with()` to construct a new value from the currently stored one.
+
+The storage operation makes use of the metric's variant of the `Metric` enumeration.
+
+## The `Metric` enumeration
+
+Persistence and in-memory serialization as well as ping payload serialization are handled through the `Metric` enumeration.
+This is defined in `glean-core/src/metrics/mod.rs`.
+Variants of this enumeration are used in the storage implementation of the metric type.
+
+To add a new metric type, add a new variant to the `Metric` enum:
+
+```rust,noplaypen
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum Metric {
+    // ...
+    Counter(i32),
+}
+```
+
+Then modify the below implementation and define the right category name for the new type. This will be used in the ping payload:
+
+```rust,noplaypen
+impl Metric {
+    pub fn category(&self) -> &'static str {
+        match self {
+            // ...
+            Metric::Counter(_) => "counter",
+        }
+    }
+}
+```
+
+Finally, define the ping payload serialization (as JSON).
+In the simple cases where the in-memory representation maps to its JSON representation it is enough to call the `json!` macro.
+
+```rust,noplaypen
+impl Metric { // same block as above
+    pub fn as_json(&self) -> JsonValue {
+        match self {
+            // ...
+            Metric::Counter(c) => json!(c),
+        }
+    }
+}
+```
+
+For more complex serialization consider implementing serialization logic as a function returning a [`serde_json::Value`](https://docs.rs/serde_json/*/serde_json/enum.Value.html)
+or another object that can be serialized.
```

Comment:
I feel like we'll want a concrete example of this, but we can defer that until we actually have one in the code base.  In particular, documenting the convention that the storage struct (as distinct from the metric definition struct) lives in the same module as the metric type etc., etc...

Reply:
Yeah, I was thinking about that but then I was half-way into implementing Datetime and stopped.
I'll file a bug so we replace this once we have a more concrete example.

Reply:
https://bugzilla.mozilla.org/show_bug.cgi?id=1552774

Reply:
Yeah, sounds good.  I think we don't fully know all the details on this until we implement the first one anyway...